### PR TITLE
Implement Scanning Parsers

### PR DESCRIPTION
**Pull request by yaauie**
The existing parsers can fail to parse correctly in a number of cases, notably:

 - when headers contain certain legal escape sequences
 - when extension values contain spaces and legal escape sequences (#54)

This change-set replaces both phases of the decode parser with scanning parsers that work with relatively-simple regular expressions. Because we are _scanning_ the input, we can handle arbitrary escape sequences without using look-behinds (which must be fixed-lenth and therefore cannot match chains of escape sequences), and we can also avoid the extra steps of munging/unmunging that the current implementation does to get escaped sequences out of the way and put them back again.

This change-set is best viewed commit-by-commit or [end-result only](https://github.com/yaauie/logstash-codec-cef/blob/scanning-parsers/lib/logstash/codecs/cef.rb), since the all-together diff is a bit of a mess.

***

*EDIT: my original PR comment included language that was unnecessarily negative; I have edited the above, but the original can be found [here](https://gist.github.com/yaauie/fca0d0f4bf8584033b7a836130e79bdd).*

### CODE REVIEW

**Comment by breml** [comment]
@yaauie can you provide some benchmark results to substantiate your statement regarding the performance?

**Comment by yaauie** [comment]
I've chased down the CI failure, which has to do with the new strict field-reference parsing (on Logstash's master branch slated for the eventual 7.0 release, available as opt-in feature starting with 6.4.0); the field-reference parser is being made more strict to better handle ambiguous input.

There are two tested inputs that are attempting to use the event API with strings that use an array-like syntax that relies on documented behaviour of the original field-reference parser.

In order to ensure that existing inputs to this codec continue to work, I'm working on a fix in this plugin that would coerce these array-syntax extension field names to be legal Logstash field-reference syntax before sending them as field names to the Logstash Event API.

I'll include the fix in this PR, even though it would be useful without changing to a streaming parser.

File: lib/logstash/codecs/cef.rb [hunk 1]
```diff
@@ -155,36 +274,21 @@ def handle(data, &block)
     # Get rid of the CEF bit in the version
     event.set('cefVersion', event.get('cefVersion').sub(/^CEF:/, ''))
 
-    # Strip any whitespace from the message
-    if not message.nil? and message.include? '='
+    # Use a scanning parser to capture the Extension Key/Value Pairs
+    if message && !message.empty? && message.include?('=')
```
Comment (colinsurprenant) [inline lib/logstash/codecs/cef.rb]: wondering if doing `!message.empty? && message.include?('=')` vs just  `message.include?('=')` in actually faster ?  - I mean if most `message` are expected to be non empty at this point I'd assume this extra `!message.empty?` call will be actually slower and wonder if for actual empty `message` checking for emptiness before calling  `include?` has a practical benefit?

File: spec/codecs/cef_spec.rb [hunk 2]
```diff
@@ -306,32 +349,45 @@ def validate(e)
...
+      let(:message) { "CEF:0|security|threatmanager|1.0|100|double backslash" +
+                      backslash + backslash + # escaped backslash
+                      backslash + backslash + # escaped backslash
+                      "|10|src=10.0.0.192 dst=12.121.122.82 spt=1232" }
 
-        expect(event).to be_truthy
+      it 'should include the backslashes unescaped' do
+        event = decode_one(subject, message)
+
+        expect(event).to be_a(LogStash::Event)
```
Comment (colinsurprenant) [inline spec/codecs/cef_spec.rb]: minor: I think this expectation is superfluous for this test. I suggest a separate test be made just to test that the decode function produces an `Event` class.

File: spec/codecs/cef_spec.rb [hunk 3]
```diff
@@ -349,35 +405,51 @@ def validate(e)
...
 
+    context('escaped-equals and unescaped-spaces in the extension values') do
+      let(:query_string) { 'key1=value1&key2=value3 aa.bc&key3=value4'}
+      let(:escaped_query_string) { query_string.gsub('=','\\=') }
+      let(:cef_message) { "CEF:0|security|threatmanager|1.0|100|trojan successfully stopped|10|go=start now query_string=#{escaped_query_string} final=done" }
+
+      it 'captures the extension values correctly' do
+        event = decode_one(subject, cef_message)
+
+        expect(event).to be_a(LogStash::Event)
```
Comment (colinsurprenant) [inline spec/codecs/cef_spec.rb]: same

**Comment by colinsurprenant** [comment]
Great refactor @yaauie. Left a few minor notes.

Since this actually improves correctness, I am +1 on merging this regardless of performance concerns **but** I would like to also see some benchmarks as @breml  suggested to make sure that if there is a significant  regression we can followup on it.

File: spec/codecs/cef_spec.rb (hunk 2 above)
Comment (yaauie) [reply spec/codecs/cef_spec.rb]: The existing `validate` function that is called on most of the rest of the specs does this before testing the contents of the event, so I was mostly copying for consistency.

You're right, though. It's not strictly necessary.

**Comment by colinsurprenant** [comment]
LGTM

**Comment by yaauie** [comment]
On my 4-core 2.9GHz laptop, using `input { stdin => { codec => cef { delimiter => "\n" } } }`, performance was as-follows:

| plugin | events per second |
| ---- | ------- |
| 5.0.3 release | ~8,800 |
| this PR | ~12,000 |

Unfortunately the corpus of data I used as input is not something I can share publicly. I sourced my corpus from a number of different places, and the inputs have a fair bit of variety in terms of the use of potentially-expensive things like escaped special characters and unescaped spaces in the extensions.

**Pull request merged**